### PR TITLE
[OM] Add om.enum type

### PR DESCRIPTION
**Pull request by uenoku**
This commit adds om.enum type.

### CODE REVIEW

File: lib/Dialect/OM/OMTypes.cpp [hunk 1]
```diff
@@ -41,3 +46,53 @@ bool circt::om::isMapKeyValuePairType(mlir::Type type) {
...
+  return llvm::hash_combine(fi.name, fi.type);
+}
+} // namespace detail
+} // namespace om
+} // namespace circt
+
+/// Parse a list of field names and types within <>. E.g.:
+/// <foo: i7, bar: i8>
+static mlir::ParseResult
+parseFields(AsmParser &p, SmallVectorImpl<EnumType::EnumElement> &parameters) {
```
Comment (prithayan) [inline lib/Dialect/OM/OMTypes.cpp:67]: Nit suggestion, 
```suggestion
parseEnumFields(AsmParser &p, SmallVectorImpl<EnumType::EnumElement> &parameters) {
```

File: lib/Dialect/OM/OMTypes.cpp [hunk 2]
```diff
@@ -41,3 +46,53 @@ bool circt::om::isMapKeyValuePairType(mlir::Type type) {
...
+        if (p.parseKeyword(&name) || p.parseColon() || p.parseType(type))
+          return failure();
+        parameters.push_back(
+            EnumType::EnumElement{StringAttr::get(p.getContext(), name), type});
+        return success();
+      });
+}
+
+/// Print out a list of named fields surrounded by <>.
+static void printFields(AsmPrinter &p, ArrayRef<EnumType::EnumElement> fields) {
```
Comment (prithayan) [inline lib/Dialect/OM/OMTypes.cpp:81]: Nit suggestion,
```suggestion
static void printEnumFields(AsmPrinter &p, ArrayRef<EnumType::EnumElement> fields) {
```

**Review by prithayan — APPROVED** [review]
Looks good, clang-tidy errors need to be fixed and feel free to ignore my suggested name changes.

**Comment by dtzSiFive** [comment]
Should this reject having multiple entries for same "key"?

Is this effectively a struct type?

**Comment by uenoku** [comment]
> Should this reject having multiple entries for same "key"?

Yes, it should. 

> Is this effectively a struct type?

No, I meant enum to be a sum type which will be used for map value type.

**Pull request merged**